File: android/mendixnative/src/main/java/com/mendix/mendixnative/react/NativeErrorHandler.java
package com.mendix.mendixnative.react;

import com.facebook.common.logging.FLog;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.ExceptionsManagerModule;

import org.jetbrains.annotations.NotNull;

// Used by previous versions of the client (<= 9.15)
@ReactModule(name = NativeErrorHandler.NAME)
public class NativeErrorHandler extends ReactContextBaseJavaModule {
    static final String NAME = "NativeErrorHandler";

    NativeErrorHandler(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NotNull
    @Override
    public String getName() {
        return NAME;
    }

    @ReactMethod
    public void handle(String message, ReadableArray stackTrace) {
        ExceptionsManagerModule exceptionsManagerModule = getReactApplicationContext().getNativeModule(ExceptionsManagerModule.class);
        exceptionsManagerModule.reportSoftException(message, stackTrace, 0);
        exceptionsManagerModule.updateExceptionMessage(message, stackTrace, 0);

        FLog.e(getClass(), "Received JS exception: " + message);
    }
}

File: android/src/main/java/com/mendix/mendixnative/react/NativeErrorHandler.kt
package com.mendix.mendixnative.react

import com.facebook.common.logging.FLog
import com.facebook.react.ReactApplication
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.devsupport.StackTraceHelper

class NativeErrorHandler(val reactContext: ReactApplicationContext) {
    fun handle(message: String?, stackTrace: ReadableArray?) {
        FLog.e(javaClass, "Received JS exception: $message")
        // In bridgeless mode, use DevSupportManager directly for proper error display
        val reactHost = (reactContext.applicationContext as? ReactApplication)?.reactHost
        reactHost?.devSupportManager?.showNewJSError(message, sanitize(stackTrace), -1)
    }

    /**
     * Filter out invalid stack frames to prevent parsing errors.
     *
     * React Native's StackTraceHelper.convertJsStackTrace() uses requireNotNull() for
     * methodName and file. Invalid frames cause secondary errors that break RedBox and reload.
     *
     * Simply skip frames that don't have the required non-null fields.
     */
    private fun sanitize(stackTrace: ReadableArray?): ReadableArray {
        val filtered = Arguments.createArray()
        if (stackTrace == null) return filtered
        (0 until stackTrace.size())
            .mapNotNull { stackTrace.getMap(it) }
            .filter { isValidFrame(it) }
            .forEach { filtered.pushMap(it) }

        return filtered
    }

    /**
     * Check if a stack frame has the required non-null fields for StackTraceHelper.
     * Uses React Native's own key constants to match their validation logic.
     */
    private fun isValidFrame(frame: ReadableMap): Boolean {
        return arrayOf(StackTraceHelper.FILE_KEY, StackTraceHelper.METHOD_NAME_KEY).all {
            frame.hasKey(it) && !frame.isNull(it)
        }
    }
}
